src/ZkFold/Cardano/UPLC/Rollup.hs

```diff
@@ -65,11 +66,9 @@ rollup (RollupSetup ledgerRules dataCurrency threadValue feeAddress) (UpdateRoll
     -- Get state updates as token names of the data currency
     update' =
       map (unTokenName . fst) $
-      concatMap (toList . snd) $
-      filter (\(k, _) -> k == dataCurrency) $
-      -- Every referenced input must have the data currency as the second currency (the first one is ada).
-      map (head . tail . toList . getValue . txOutValue . txInInfoResolved)
-      (txInfoReferenceInputs $ scriptContextTxInfo ctx)
+      concatMap toList $
```

Comment:
Using `lookup` idea instead of `filter` here.  This appears to be the right way of dealing with the fact that we have an extra *reference input* due to the presence of a *reference script*.  We needed to account for the fact that this extra reference inputs does not have the value for the data update, and we can not even control in which order it appears in the `txInfoReferenceInputs` list.  Usage of `lookup` solves all this.

Reply:
Agree